sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,443 @@
+pub const MAX_SIZE: usize = NUM_ENTRIES_SIZE + MAX_ENTRIES * ENTRY_SIZE;
+
+/// A single entry in the `SlotHashes` sysvar.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(C)]
+pub struct SlotHashEntry {
+    /// The slot number stored as little-endian bytes.
+    slot_le: [u8; 8],
+    /// The hash corresponding to the slot.
+    pub hash: [u8; HASH_BYTES],
+}
+
+const _: [(); 1] = [(); mem::align_of::<SlotHashEntry>()];
+
+/// SlotHashes provides read-only, zero-copy access to SlotHashes sysvar bytes.
+pub struct SlotHashes<T: Deref<Target = [u8]>> {
+    data: T,
+    /// Number of entries (decoded from the 8-byte prefix).  Immutable.
+    len: usize,
+}
+
+/// Reads the entry count from the first 8 bytes of data.
+/// Returns None if the data is too short.
+#[inline(always)]
+pub(crate) fn read_entry_count_from_bytes(data: &[u8]) -> Option<usize> {
+    if data.len() < NUM_ENTRIES_SIZE {
+        return None;
+    }
+    Some(unsafe { u64::from_le_bytes(*(data.as_ptr() as *const [u8; 8])) } as usize)
+}
+
+/// Reads the entry count from the first 8 bytes of data.
+///
+/// # Safety
+/// Caller must ensure data has at least NUM_ENTRIES_SIZE bytes.
+#[inline(always)]
+pub(crate) unsafe fn read_entry_count_from_bytes_unchecked(data: &[u8]) -> usize {
+    u64::from_le_bytes(*(data.as_ptr() as *const [u8; 8])) as usize
+}
+
+/// Validates core SlotHashes constraints: entry count and buffer size requirements.
+///
+/// # Arguments
+/// * `buffer_len` - Total buffer length including 8-byte header
+/// * `declared_entries` - Optional declared entry count from header (None to skip this check)
+///
+/// # Returns
+/// The maximum entries that fit in the buffer, or error if constraints violated
+#[inline]
+fn validate_slothashes_constraints(
+    buffer_len: usize,
+    declared_entries: Option<usize>,
+) -> Result<usize, ProgramError> {
+    // Must have space for 8-byte header
+    if buffer_len < NUM_ENTRIES_SIZE {
+        return Err(ProgramError::AccountDataTooSmall);
+    }
+
+    // Calculate how many entries can fit
+    let data_len = buffer_len - NUM_ENTRIES_SIZE;
+    if data_len % ENTRY_SIZE != 0 {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let max_entries = data_len / ENTRY_SIZE;
+    if max_entries > MAX_ENTRIES {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    if let Some(declared) = declared_entries {
+        if declared > max_entries {
+            return Err(ProgramError::InvalidArgument);
+        }
+        return Ok(declared);
+    }
+
+    Ok(max_entries)
+}
+
+/// Validates SlotHashes data format and returns the entry count.
+#[inline]
+fn parse_and_validate_data(data: &[u8]) -> Result<usize, ProgramError> {
+    // Need at least the 8-byte length prefix.
+    if data.len() < NUM_ENTRIES_SIZE {
+        return Err(ProgramError::AccountDataTooSmall);
+    }
+
+    let num_entries = unsafe { read_entry_count_from_bytes_unchecked(data) };
```

Comment:
*minor*

`read_entry_count_from_bytes()` performs the length check, so if you use it, it would allow you to avoid an unsafe block:

```suggestion
    let num_entries = read_entry_count_from_bytes(data)
        .ok_or(ProgramError::AccountDataTooSmall)?;
```

Reply:
Nice catch, addressed in d422291c5af5263155b5412b663c347478ab3258